v-next/hardhat/src/internal/builtin-plugins/console/task-action.ts

```diff
@@ -0,0 +1,70 @@
+import type { NewTaskActionFunction } from "@ignored/hardhat-vnext-core/types/tasks";
+import type { REPLServer } from "node:repl";
+
+import path from "node:path";
+import repl from "node:repl";
+
+import { getCacheDir } from "@ignored/hardhat-vnext-core/global-dir";
+import debug from "debug";
+
+const log = debug("hardhat:core:tasks:console");
+
+interface ConsoleActionArguments {
+  commands: string[];
+  history: string;
+  noCompile: boolean;
+  // We accept ReplOptions as an argument to allow tests overriding the IO streams
+  options?: repl.ReplOptions;
+}
+
+const consoleAction: NewTaskActionFunction<ConsoleActionArguments> = async (
+  { commands, history, noCompile, options },
+  hre,
+) => {
+  // Resolve the history path if it is not empty
+  let historyPath: string | undefined;
+  if (history !== "") {
+    const globalCacheDir = await getCacheDir();
+    historyPath = path.isAbsolute(history)
+      ? history
+      : path.resolve(globalCacheDir, history);
+  }
+
+  // If noCompile is false, run the compile task first
+  if (!noCompile) {
+    // todo: run compile task
+  }
+
+  return new Promise<REPLServer>(async (resolve) => {
+    // Start a new REPL server with the default options
+    const replServer = repl.start(options);
+
+    // Resolve the task action promise only when the REPL server exits
+    replServer.on("exit", () => {
+      resolve(replServer);
+    });
+
+    // Add the Hardhat Runtime Environment to the REPL context
+    replServer.context.hre = hre;
```

Comment:
@alcuadrado I'm continuing [the dicsussion about the `hre` in the context](https://github.com/NomicFoundation/hardhat/pull/5592#discussion_r1712244971) here since I now added it.

> I think what's missing here is exporting the hre. I believe that's done with replServer.context.hre = here.
> 
> We should also consider exposing each hre field in the context, similar to what we do in v-next/hardhat/src/index.ts

I didn't expose it initially because I assumed we wanted the users to import it explicitly with `const { default: hre } = await import("@ignored/hardhat-vnext");` for example.

Here's also a suggestion for exposing other hre fields in the context. Let me know if we should proceed with it.   

```suggestion
    // Add the Hardhat Runtime Environment to the REPL context
    replServer.context.hre = hre;
    replServer.context.config = config;
    replServer.context.tasks = tasks;
    replServer.context.globalOptions = globalOptions;
    replServer.context.hooks = hooks;
    replServer.context.interruptions = interruptions;
```

Reply:
> I didn't expose it initially because I assumed we wanted the users to import it explicitly with const { default: hre } = await import("@ignored/hardhat-vnext"); for example.

We want that for tests/scripts, but I think in the console it makes sense to automatically import them, similarly to what node's own console does.

Reply:
Let's commit that change, yes